### PR TITLE
UI changes in Add Metrics page

### PR DESCRIPTION
**Pull request by KD1712**
Replaced ID column with Context column, added context filter option. In ADD METRIC json-editor, added context list and removed highlighted blue line

### CODE REVIEW

**Comment by KD1712** [comment]
@zackcl @danoswaltCL  We are updating the UI of our table component, which includes a hierarchical tree structure. Initially, we used drop-down icons for all nodes. Now, changes made here to create difference between leaf node and expandable are: Expandable nodes use the drop-down icon and leaf nodes use bullet points (•).

Which design option do we prefer for representing leaf nodes in our table component:
**Bullet Points** OR **Empty Space** OR **Disabled drop-down icon** ?

**Review by ppratikcr7 — CHANGES_REQUESTED** [review]
I found the following things in the UI:

1. The chips are not spaced as they should be in tags chips on experiment page.

2. The bullet points are square, can you make them round bullets? @zackcl What is your opinion on this?

3. The context chips when clicked, wont trigger search operation as we have on experiments page.

4. Can the alignment of the bullet point and arrow for different rows be more aligned in a straight line?
Screenshot 2024-06-03 at 7 37 43 PM
Screenshot 2024-06-03 at 7 37 43 PM

**Comment by hsnee** [comment]
FYI @zackcl is out until July 1st. Tagging @amurphy-cl to decide if someone else should have input on these.

**Comment by amurphy-cl** [comment]
Thanks Husni, @danoswaltCL can you have a look?

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts [hunk 1]
```diff
@@ -11,10 +14,18 @@ import { AnalysisService } from '../../../../../../core/analysis/analysis.servic
...
+  allMetricsSub: Subscription;
+  contextOptions: string[] = [];
   @ViewChild('metricsEditor', { static: false }) metricsEditor: JsonEditorComponent;
   constructor(private dialogRef: MatDialogRef<AddMetricsComponent>, private analysisService: AnalysisService) {}
 
   ngOnInit() {
+    this.allMetricsSub = this.analysisService.allMetrics$.subscribe((metrics: IMetricUnit[]) => {
+      this.allMetrics = metrics;
+      this.extractContext(this.allMetrics);
+      console.log(this.allMetrics);
```
Comment (RidhamShah) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts:27]: remove this console.log

**Comment by ppratikcr7** [comment]
> I found the following things in the UI:
> 
> 1. The chips are not spaced as they should be in tags chips on experiment page.
> 2. The bullet points are square, can you make them round bullets? @zackcl What is your opinion on this?
> 3. The context chips when clicked, wont trigger search operation as we have on experiments page.
> 4. Can the alignment of the bullet point and arrow for different rows be more aligned in a straight line?
> 
> Screenshot 2024-06-03 at 7 37 43 PM Screenshot 2024-06-03 at 7 37 56 PM

Adding one more thing, the context from dropdown on json editor is not saving metrics successfully, I guess that hook with backend is not implemented yet, and context might be needed in the JSON itself. I am using the below JSON to add new metrics:

`{
  "metrics": [
       {
            "metric": "totalTimeSeconds3", 
            "datatype": "continuous"
       }
  ]
}`

**Comment by KD1712** [comment]
> >  @danoswaltCL We are updating the UI of our table component, which includes a hierarchical tree structure. Initially, we used drop-down icons for all nodes. Now, changes made here to create difference between leaf node and expandable are: Expandable nodes use the drop-down icon and leaf nodes use bullet points (•).
> > Which design option do we prefer for representing leaf nodes in our table component: **Bullet Points** OR **Empty Space** OR **Disabled drop-down icon** ?
> 
> > @danoswaltCL We are updating the UI of our table component, which includes a hierarchical tree structure. Initially, we used drop-down icons for all nodes. Now, changes made here to create difference between leaf node and expandable are: Expandable nodes use the drop-down icon and leaf nodes use bullet points (•).
> > Which design option do we prefer for representing leaf nodes in our table component: **Bullet Points** OR **Empty Space** OR **Disabled drop-down icon** ?
> 
> @KD1712 Can you please add the reference image?

Screenshot 2024-06-05 at 7 36 46 PM

**Comment by ppratikcr7** [comment]
@KD1712 Thanks for adding a padding between the text and delete icon. I noticed one more thing, the delete icon is black, but we use a grey shade at all other places. Can you pick that shade from design stepper page?
Screenshot 2024-06-06 at 5 45 20 PM

**Comment by ppratikcr7** [comment]
@KD1712 I checked again, all looks good now. But, the delete button in the delete metrics modal is not enabled on correct metric path name. It is was working earlier and works in dev branch as well. Something is changed in this PR. Can you please check?

File: frontend/projects/upgrade/src/app/core/experiment-design-stepper/experiment-design-stepper.service.ts [hunk 2]
```diff
@@ -338,7 +338,7 @@ export class ExperimentDesignStepperService {
     const payloadTableData = this.getSimpleExperimentPayloadTableData();
 
     payloadTableData.forEach((payloadRowData: SimpleExperimentPayloadTableRowData) => {
-      const parentCondition = conditions.find((condition) => condition.conditionCode === payloadRowData.condition).id;
+      const parentCondition = conditions.find((condition) => condition.conditionCode === payloadRowData.condition);
```
Comment (danoswaltCL) [inline frontend/projects/upgrade/src/app/core/experiment-design-stepper/experiment-design-stepper.service.ts:341]: this change seems unintentional

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts [hunk 3]
```diff
@@ -112,17 +114,32 @@ export class MetricsComponent implements OnInit, OnDestroy, AfterViewInit {
...
   }
 
   applyFilter(filterValue: string) {
-    this.selectedMetricIndex = null;
-    this.analysisService.setMetricsFilterValue(filterValue);
+    this.filterMetricsPredicate(this.selectedMetricFilterOption);
+    if (typeof filterValue === 'string') {
+      this.allMetrics.filter = filterValue.trim().toLowerCase();
+    }
+  }
```
Comment (ppratikcr7) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts:126]: Add a line space between functions

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts [hunk 4]
```diff
@@ -112,17 +114,32 @@ export class MetricsComponent implements OnInit, OnDestroy, AfterViewInit {
...
+          );
+        case METRIC_SEARCH_KEY.NAME:
+          return filter !== METRIC_SEARCH_KEY.NAME ? data.key.toLocaleLowerCase().includes(filter) : data.key;
+
+        case METRIC_SEARCH_KEY.CONTEXT:
+          return !!data.context.filter((context) => context.toLocaleLowerCase().includes(filter)).length;
+      }
+    };
   }
-
```
Comment (ppratikcr7) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/metrics/metrics.component.ts:125]: Add a line space between functions

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts [hunk 5]
```diff
@@ -1,20 +1,44 @@
-import { Component, OnInit, ViewChild } from '@angular/core';
+import { Component, OnDestroy, OnInit, ViewChild } from '@angular/core';
 import { JsonEditorOptions, JsonEditorComponent } from 'ang-jsoneditor';
 import { MatDialogRef } from '@angular/material/dialog';
 import { AnalysisService } from '../../../../../../core/analysis/analysis.service';
+import { Subscription } from 'rxjs';
+// import { MetricUnit } from '../../../../../../core/analysis/store/analysis.models';
```
Comment (ppratikcr7) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts:6]: remove this commented import

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts [hunk 6]
```diff
@@ -1,20 +1,44 @@
...
+
```
Comment (ppratikcr7) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts:7]: remove this blank line between imports

File: frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts [hunk 7]
```diff
@@ -1,20 +1,44 @@
...
   styleUrls: ['./add-metrics.component.scss'],
 })
-export class AddMetricsComponent implements OnInit {
+export class AddMetricsComponent implements OnInit, OnDestroy {
   options = new JsonEditorOptions();
   metricsEditorError = false;
+
+  contextMetaData: IContextMetaData | Record<string, unknown> = {};
+  contextMetaDataSub: Subscription;
+  allContexts = [];
```
Comment (ppratikcr7) [inline frontend/projects/upgrade/src/app/features/dashboard/profile/components/modals/add-metrics/add-metrics.component.ts:22]: can you please provide the datatype for this?

**Review by ppratikcr7 — CHANGES_REQUESTED** [review]
@KD1712 Have few comments, please check.

File: frontend/projects/upgrade/src/app/core/experiment-design-stepper/experiment-design-stepper.service.ts (hunk 2 above)
Comment (ppratikcr7) [reply frontend/projects/upgrade/src/app/core/experiment-design-stepper/experiment-design-stepper.service.ts:341]: Yeah, its the same ghost change from auto merge for which I had created 2 PR's which allowed creation of experiments. I pointed out the same in my below review as well. :)

**Review by ppratikcr7 — APPROVED** [review]

**Review by ppratikcr7 — APPROVED** [review]

**Pull request merged**